mypy/checkmember.py

```diff
@@ -96,155 +96,179 @@ def analyze_member_access(name: str,
+            #
+            # This check makes sure that when we encounter an operator, we skip looking up
+            # the corresponding method in the current instance to avoid this edge case.
+            # See https://github.com/python/mypy/pull/1787 for more info.
+            result = analyze_class_attribute_access(ret_type, name, mx)
             if result:
-                if not (isinstance(result, AnyType) and item.type.fallback_to_any):
-                    return result
-                else:
-                    # We don't want errors on metaclass lookup for classes with Any fallback
-                    mx = mx.copy_modified(messages=ignore_messages)
-        if item is not None:
-            fallback = item.type.metaclass_type or fallback
+                return result
+        # Look up from the 'type' type.
+        return _analyze_member_access(name, typ.fallback, mx)
+    else:
+        assert False, 'Unexpected type {}'.format(repr(ret_type))
+
+
+def analyze_type_type_member_access(name: str, typ: TypeType, mx: MemberContext) -> Type:
+    # Similar to analyze_type_callable_attribute_access.
+    item = None
+    fallback = mx.builtin_type('builtins.type')
+    ignore_messages = mx.msg.copy()
+    ignore_messages.disable_errors()
+    if isinstance(typ.item, Instance):
+        item = typ.item
+    elif isinstance(typ.item, AnyType):
+        mx = mx.copy_modified(messages=ignore_messages)
         return _analyze_member_access(name, fallback, mx)
+    elif isinstance(typ.item, TypeVarType):
+        if isinstance(typ.item.upper_bound, Instance):
+            item = typ.item.upper_bound
+    elif isinstance(typ.item, TupleType):
+        item = typ.item.fallback
+    elif isinstance(typ.item, FunctionLike) and typ.item.is_type_obj():
+        item = typ.item.fallback
+    elif isinstance(typ.item, TypeType):
+        # Access member on metaclass object via Type[Type[C]]
+        if isinstance(typ.item.item, Instance):
+            item = typ.item.item.type.metaclass_type
+    if item and not mx.is_operator:
+        # See comment above for why operators are skipped
+        result = analyze_class_attribute_access(item, name, mx)
+        if result:
+            if not (isinstance(result, AnyType) and item.type.fallback_to_any):
+                return result
+            else:
+                # We don't want errors on metaclass lookup for classes with Any fallback
+                mx = mx.copy_modified(messages=ignore_messages)
+    if item is not None:
+        fallback = item.type.metaclass_type or fallback
+    return _analyze_member_access(name, fallback, mx)
 
+
+def analyze_union_member_access(name: str, typ: UnionType, mx: MemberContext) -> Type:
+    mx.msg.disable_type_names += 1
+    results = [_analyze_member_access(name, subtype, mx)
+               for subtype in typ.relevant_items()]
+    mx.msg.disable_type_names -= 1
+    return UnionType.make_simplified_union(results)
```

Comment:
This one is probably to small to have its own function, but for the sake of consistency it is probably OK.